Sqlite: add more locking to DBLayer methods

rvl opened the pull request:
Relates to issue #154.

It seems to be best to put a big lock around all operations which use the SQLite connection.

I was getting the following failed test:

```
  λ> quickCheck (prop_parallel db)
  *** Failed! Exception: 'SQLite3 returned ErrorError while attempting to perform step: not an error' (after 5 tests and 3 shrinks):
  ParallelCommands
    { prefix = Commands { unCommands = [] }
    , suffixes =
        [ Pair
            { proj1 =
                Commands
                  { unCommands =
                      [ Command
                          (At (ReadTxHistory (Val (MWid "c"))))
                          (At (Resp (Right (TxHistory (fromList [])))))
                          []
                      ]
                  }
            , proj2 =
                Commands
                  { unCommands =
                      [ Command
                          (At (PutTxHistory (Val (MWid "a")) (fromList [])))
                          (At (Resp (Left (NoSuchWallet (Reference (Symbolic (Var 0)))))))
                          [ Var 0 ]
                      ]
                  }
            }
        ]
    }
```

uroboros commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -188,10 +188,17 @@ updateWallet :: MonadIO m
-                 Kernel.updateHdWallet wallet rootId newLevel newName
-    updateSyncState wallet rootId v1wal
+    db <- liftIO $ Kernel.getWalletSnapshot wallet
+    case addressPoolGapByRootId rootId db of
```
We are using `addressPoolGapByRootId` for asking "is EO wallet". I think it's worth making it more explicit with a wrapper `isEOWallet :: DB -> HdRootId -> Bool`

uroboros replied:
(we might as well add an isFOWallet too?)

denisshevchenko replied:
Agree. I'll do it.

uroboros commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -204,22 +211,29 @@ updateEosWallet
-                []    -> Right $ toEosWallet db newHdRoot newGap
-                (e:_) -> Left e
+    db' <- liftIO $ Kernel.getWalletSnapshot wallet
+    case addressPoolGapByRootId rootId db' of
```
if you use `eosAccountsByRootId` here instead of `addressPoolGapByRootId`, this function will simplify quite a bit

uroboros commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -204,22 +210,27 @@ updateEosWallet
-            exceptT $ case lefts result of
-                []    -> Right $ toEosWallet db newHdRoot newGap
-                (e:_) -> Left e
+    db' <- liftIO $ Kernel.getWalletSnapshot wallet
```
As it stands, we're repeating logic that is captured by `eosAccountsByRootId`.
If you use that function instead of `getWalletOwnership` you'll get
* accountIds, no need to read them again
* checks for `GetEosWalletErrorNoAccounts` and `GetEosWalletErrorWrongAccounts` are built in

akegalj replied:
@denisshevchenko what are your thoughts on this?

denisshevchenko replied:
Yes, good idea. I'll use that function.

denisshevchenko replied:
Done.

uroboros commented on src/Cardano/Wallet/WalletLayer/Kernel/Wallets.hs:
```diff
@@ -188,10 +189,15 @@ updateWallet :: MonadIO m
-                 Kernel.updateHdWallet wallet rootId newLevel newName
-    updateSyncState wallet rootId v1wal
+    db <- liftIO $ Kernel.getWalletSnapshot wallet
+    case getWalletOwnership rootId db of
```
A (debatable) suggestion...

We can simplify these calls to `getWalletOwnership` with a helper function, e.g. 

`ensureOwnership :: WalletOwnership -> (HdRootId -> e) -> ...`

and use it like this:

_ <- ensureOwnership WalletFullyOwned (UpdateWalletError . HD.UnknownHdRoot)
-- from here on we can assume correct ownership...

akegalj replied:
I like the proposal but underlying code would not change much. We would still have to do pattern on `case` expression - only some bits of the code would be rearanged slightly.

But its up to you - I am fine with both solutions

rvl committed:
Sqlite: add more locking to DBLayer methods

It seems to be best to put a big lock around all operations which use
the SQLite connection.

I was getting the following failed test:

  λ> quickCheck (prop_parallel db)
  *** Failed! Exception: 'SQLite3 returned ErrorError while attempting to perform step: not an error' (after 5 tests and 3 shrinks):
  ParallelCommands
    { prefix = Commands { unCommands = [] }
    , suffixes =
        [ Pair
            { proj1 =
                Commands
                  { unCommands =
                      [ Command
                          (At (ReadTxHistory (Val (MWid "c"))))
                          (At (Resp (Right (TxHistory (fromList [])))))
                          []
                      ]
                  }
            , proj2 =
                Commands
                  { unCommands =
                      [ Command
                          (At (PutTxHistory (Val (MWid "a")) (fromList [])))
                          (At (Resp (Left (NoSuchWallet (Reference (Symbolic (Var 0)))))))
                          [ Var 0 ]
                      ]
                  }
            }
        ]
    }
